source/common/network/listener_filter_buffer_impl.cc

```diff
@@ -0,0 +1,105 @@
+#include "source/common/network/listener_filter_buffer_impl.h"
+
+#include <string>
+
+namespace Envoy {
+namespace Network {
+
+const Buffer::ConstRawSlice ListenerFilterBufferImpl::rawSlice() const {
+  Buffer::ConstRawSlice slice;
+  auto front_slice = buffer_->frontSlice();
+  slice.mem_ = static_cast<uint8_t*>(front_slice.mem_) + drained_size_;
+  slice.len_ = data_size_;
+  return slice;
+}
+
+uint64_t ListenerFilterBufferImpl::drain(uint64_t length) {
+  ASSERT(length <= data_size_);
+  auto size_to_drain = std::min(length, data_size_);
+  // It doesn't drain the data from buffer directly until drain the data
```

Comment:
Why is this a separate step from `drainFromSocket`? Why not do it here?

Reply:
emm...good question, I thought I didn't want to the filter to deal with the socket error, and leave those to the framework. But second thought, there should be no any problem, I can just return boolean indicate the drain successful or not. If drain failed, the listener can close the connection and stop the filters iteration.

Reply:
Yes, I think if the drain fails, closing/aborting the connection is fine.